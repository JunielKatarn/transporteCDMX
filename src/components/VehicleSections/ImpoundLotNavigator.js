import React, { PureComponent, Fragment } from 'react'
import { Dimensions, StyleSheet } from 'react-native'
import { Snackbar } from 'react-native-paper'
import NavigationRow from '../NavigationRow'
import { translate } from '../../i18n'
import { getImpoundLots } from '../../api/endpoints'
import ImpoundLot from '../../models/ImpoundLot'
import { HTTP_SUCCESS } from '../../api/request'
// TODO: This should come from the server  ¯\_(ツ)_/¯
const RELEASE_CODE = 2

const codeDetail = {
  0: translate('impound_lot_in_transit'),
  1: translate('impound_lot_arrived'),
  2: translate('impound_lot_delivered'),
  3: translate('impound_lot_change'),
}

class ImpoundLotNavigator extends PureComponent {
  state = {
    impoundLots: null,
    isLoading: true,
    error: false,
    errorVisibility: false,
  }

  _impoundLotsResponseSuccess = impoundLotsResponse => {
    if (impoundLotsResponse?.data?.length) {
      const impoundLots = impoundLotsResponse.data.map(impoundLot => {
        return new ImpoundLot(impoundLot)
      })
      this.setState({ impoundLots, error: false })
    } else {
      this.setState({ impoundLots: null, error: true, errorVisibility: true })
    }
  }

  _impoundResponseHandler = async impoundsResponse => {
    if (impoundsResponse.status === HTTP_SUCCESS) {
      this._impoundLotsResponseSuccess(impoundsResponse)
    } else {
      this.setState({ impoundLots: null, error: true, errorVisibility: true })
    }
  }

  _getNavigationDetailLabel = () => {
    if (this.state.isLoading) {
      return translate('loading')
    }

    if (this.state.error) {
      return translate('error')
    }

    if (this.state.impoundLots?.length) {
      if (this.state.impoundLots[this.state.impoundLots.length - 1].code !== RELEASE_CODE)
        return translate('impounded')
    }

    return translate('not_impounded')
  }

  _unSetErrorVisibility = () => {
    this.setState({ errorVisibility: false })
  }

  getImpoundLotsInfo = async () => {
    if (this.props.plateText) {
      try {
        this.setState({ isLoading: true, error: false, errorVisibility: false })
        this._impoundResponseHandler(await getImpoundLots(this.props.plateText))
        this.setState({ isLoading: false })
      } catch (e) {
        this.setState({ isLoading: false, error: true, errorVisibility: true })
        console.log(e)
      }
    }
  }

  render() {
    return (
      <Fragment>
        <NavigationRow
          onPress={this.state.error ? null : () => null}
          iconName="car"
          iconColor="cornflowerblue"
          navigationDetail={this._getNavigationDetailLabel()}
          text={translate('impound_lot')}
          isLoading={this.state.isLoading}
        />
        <Snackbar
          visible={this.state.errorVisibility}
          onDismiss={this._unSetErrorVisibility}
          style={styles.snackbar}>
          {translate('there_was_an_error')}
        </Snackbar>
      </Fragment>
    )
  }
}

const styles = StyleSheet.create({
  snackbar: {
    position: 'absolute',
    bottom: -Dimensions.get('window').height / 2.4,
  },
})

export default ImpoundLotNavigator
