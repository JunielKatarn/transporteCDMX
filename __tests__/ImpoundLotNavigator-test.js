import React from 'react'
import { shallow } from 'enzyme'
import ImpoundLotNavigator from '../src/components/VehicleSections/ImpoundLotNavigator'
import impoundLots from './fixtures/impoundLots-fixture'

describe('ImpoundLotNavigator', () => {
  it('it renders correctly when NOT impounded', () => {
    const wrapper = shallow(<ImpoundLotNavigator />)
    wrapper.setState({ impoundLots, isLoading: false })
    expect(wrapper).toMatchSnapshot()
  })

  it('it renders correctly when impounded', () => {
    const wrapper = shallow(<ImpoundLotNavigator />)
    wrapper.setState({ impoundLots: [impoundLots[0]], isLoading: false })
    expect(wrapper).toMatchSnapshot()
  })

  it('it renders correctly when it has an error', () => {
    const wrapper = shallow(<ImpoundLotNavigator />)
    wrapper.setState({ impoundLots: null, isLoading: false, error: true })
    expect(wrapper).toMatchSnapshot()
  })

  it('it renders correctly when is loading', () => {
    const wrapper = shallow(<ImpoundLotNavigator />)
    wrapper.setState({ impoundLots: null, isLoading: true, error: false })
    expect(wrapper).toMatchSnapshot()
  })
})
