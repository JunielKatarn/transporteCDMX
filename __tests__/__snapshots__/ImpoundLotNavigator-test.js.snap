// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImpoundLotNavigator it renders correctly when NOT impounded 1`] = `
<Fragment>
  <Memo(NavigationRow)
    iconColor="cornflowerblue"
    iconName="car"
    isLoading={false}
    navigationDetail="[missing \\"en.not_impounded\\" translation]"
    onPress={[Function]}
    text="[missing \\"en.impound_lot\\" translation]"
  />
</Fragment>
`;

exports[`ImpoundLotNavigator it renders correctly when impounded 1`] = `
<Fragment>
  <Memo(NavigationRow)
    iconColor="cornflowerblue"
    iconName="car"
    isLoading={false}
    navigationDetail="[missing \\"en.impounded\\" translation]"
    onPress={[Function]}
    text="[missing \\"en.impound_lot\\" translation]"
  />
</Fragment>
`;

exports[`ImpoundLotNavigator it renders correctly when is loading 1`] = `
<Fragment>
  <Memo(NavigationRow)
    iconColor="cornflowerblue"
    iconName="car"
    isLoading={true}
    navigationDetail="[missing \\"en.loading\\" translation]"
    onPress={[Function]}
    text="[missing \\"en.impound_lot\\" translation]"
  />
</Fragment>
`;

exports[`ImpoundLotNavigator it renders correctly when it has an error 1`] = `
<Fragment>
  <Memo(NavigationRow)
    iconColor="cornflowerblue"
    iconName="car"
    isLoading={false}
    navigationDetail="[missing \\"en.error\\" translation]"
    onPress={null}
    text="[missing \\"en.impound_lot\\" translation]"
  />
</Fragment>
`;
